gstd/src/common/handlers.rs

```diff
@@ -25,50 +25,208 @@
+/// has not changed:
+/// ```
+/// use std::panic::{self, PanicInfo};
+///
+/// const MESSAGE: &str = "message";
+///
+/// fn expected_format(panic_info: &PanicInfo) -> String {
+///     let location = panic_info.location().unwrap();
+///     format!("panicked at {location}:\n{MESSAGE}")
+/// }
+///
+/// panic::set_hook(Box::new(|panic_info| {
+///     assert_eq!(panic_info.to_string(), expected_format(panic_info));
+/// }));
+///
+/// let result = panic::catch_unwind(|| {
+///     panic!("{MESSAGE}");
+/// });
+/// assert!(result.is_err());
+/// ```
+pub mod panic_handler {
+    #[cfg(target_arch = "wasm32")]
+    mod internal {
+        use crate::ext;
+        use core::panic::PanicInfo;
 
-    #[cfg(not(feature = "debug"))]
-    #[panic_handler]
-    pub fn panic(_: &PanicInfo) -> ! {
-        ext::panic("no info")
-    }
+        /// Minimal panic handler.
+        #[cfg(not(any(feature = "panic-message", feature = "panic-location")))]
+        #[panic_handler]
+        pub fn panic(_: &PanicInfo) -> ! {
+            ext::panic("no info")
+        }
+
+        #[cfg(any(feature = "panic-message", feature = "panic-location"))]
+        mod constants {
+            /// Max amount of bytes allowed to be thrown as string explanation
+            /// of the error.
+            pub const TRIMMED_MAX_LEN: usize = 1024; //TODO: do not duplicate `gear_core::str::TRIMMED_MAX_LEN`
+            /// This prefix is used to print debug message:
+            /// `debug!("panic occurred: {msg}")`.
+            pub const PANIC_OCCURRED: &str = "panic occurred: ";
+            /// Size of array string that will be allocated on the stack.
+            pub const ARRAY_STRING_MAX_LEN: usize = if cfg!(feature = "debug") {
+                PANIC_OCCURRED.len()
+            } else {
+                0
+            } + TRIMMED_MAX_LEN;
+            /// This prefix is used by `impl Display for PanicInfo<'_>`.
+            #[cfg(not(feature = "nightly"))]
+            pub const PANICKED_AT: &str = "panicked at ";
+        }
+
+        #[cfg(any(feature = "panic-message", feature = "panic-location"))]
+        use constants::*;
+
+        /// Panic handler for nightly Rust.
+        #[cfg(feature = "nightly")]
+        #[cfg(any(feature = "panic-message", feature = "panic-location"))]
+        #[panic_handler]
+        pub fn panic(panic_info: &PanicInfo) -> ! {
+            use crate::prelude::fmt::Write;
+            use arrayvec::ArrayString;
+
+            let mut debug_msg = ArrayString::<ARRAY_STRING_MAX_LEN>::new();
+
+            #[cfg(feature = "debug")]
+            let _ = debug_msg.try_push_str(PANIC_OCCURRED);
+
+            let _ = match (panic_info.message(), panic_info.location()) {
+                #[cfg(all(feature = "panic-message", feature = "panic-location"))]
+                (Some(msg), Some(loc)) => write!(&mut debug_msg, "'{msg}', {loc}"),
+                #[cfg(all(feature = "panic-message", not(feature = "panic-location")))]
```

Comment:
just match cfg(feat = location), or cfg(not(feat = location))

Reply:
9d99503e38b4d15e7f77d27eff6da71ec6013f86